pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -434,3 +486,89 @@ func (s *shardActiveSeriesMiddleware) writeMergedResponse(ctx context.Context, c
 
 	stream.WriteObjectEnd()
 }
+
+func (s *shardActiveSeriesMiddleware) writeMergedResponseWithZeroAllocationDecoder(ctx context.Context, check func() error, w io.WriteCloser, items chan chan activeSeriesDataChunk, encoding string) {
+	defer w.Close()
+
+	span, _ := opentracing.StartSpanFromContext(ctx, "shardActiveSeries.writeMergedResponseWithZeroAllocationDecoder")
+	defer span.Finish()
+
+	var out io.Writer = w
+	if encoding == encodingTypeSnappyFramed {
+		span.LogFields(otlog.String("encoding", encodingTypeSnappyFramed))
+		enc := getSnappyWriter(w)
+		out = enc
+		defer func() {
+			enc.Close()
+			// Reset the encoder before putting it back to pool to avoid it to hold the writer.
+			enc.Reset(nil)
+			snappyWriterPool.Put(enc)
+		}()
+	} else {
+		span.LogFields(otlog.String("encoding", "none"))
+	}
+
+	stream := jsoniter.ConfigFastest.BorrowStream(out)
+	defer func(stream *jsoniter.Stream) {
+		_ = stream.Flush()
+
+		if cap(stream.Buffer()) > jsoniterMaxBufferSize {
+			return
+		}
+		jsoniter.ConfigFastest.ReturnStream(stream)
+	}(stream)
+
+	stream.WriteObjectStart()
+	stream.WriteObjectField("data")
+	stream.WriteArrayStart()
+	firstItem := true
+	for chunkCh := range items {
+		chunk := <-chunkCh
+		if chunk.done {
+			continue
+		}
+
+		if firstItem {
+			firstItem = false
+		} else {
+			stream.WriteMore()
+		}
+
+		for {
+			chunkStr := unsafe.String(
+				unsafe.SliceData(chunk.buff.Bytes()),
+				chunk.buff.Len(),
+			)
+
+			// Write the value as is, since it's already a JSON array.
+			stream.WriteRaw(chunkStr)
```

Comment:
Any reason not to use `Write()` here and skip the unsafe string conversion?

Reply:
main reason is that by using of `*stream.Write` method, which also invokes an underlying `io.Writer` interface, forces the compiler to escape the `buf.Bytes()` slice to the heap, due to the fact that the compiler has no way of knowing whether or not the slice is retained by the implementation. anyway, this does not make a huge difference in total number of allocations, so if you consider it's worth making the sacrifice, we can make the change.